src/librustc_typeck/check/pat.rs

```diff
@@ -380,9 +380,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 E0029,
                 "only char and numeric types are allowed in range patterns"
             );
-            err.span_label(span, "ranges require char or numeric types");
-            err.note(&format!("start type: {}", self.ty_to_string(lhs_ty)));
-            err.note(&format!("end type: {}", self.ty_to_string(rhs_ty)));
+            if !lhs_compat && !rhs_compat {
```

Comment:
At this point we have so many negations that it might be more readable to bake the negation into `lhs/rhs_compat` themselves and rename `lhs_fail`.